Pass ForwardOptions from top level module and also return any relevant state as output

sxu opened the pull request:
Summary: Pass a `ForwardOptions` argument (introduced by https://github.com/pytorch/executorch/pull/8128) from the top level transformer, consolidate some existing inputs into it, and return any optional updates from the attention implementation. This is in preparation for a static shape attention implementation.

Differential Revision: D69080123




cc @mergennachin @cccclai @helunwencser @dvorjackz

iseeyuan commented:
@pytorchbot label "release notes: llama_transformer: Pass ForwardOptions from top level module and also return any relevant state as output"

cccclai commented on examples/models/llama/attention.py:
```diff
@@ -236,7 +236,7 @@ def forward(
             k, v = self.kv_cache.update(input_pos, k, v)
             output = self.SDPA(input_pos, q, k, v, bsz, seqlen, self.mask)
-            return self.wo(output)
+            return self.wo(output), None
```
Is `None` a place holder or something else?

sxu replied:
It's to support additional output from each attention layer that'll be aggregated and returned by the top level transformer. To avoid disruptions if this is `None, the top level transformer still return a single logit output, otherwise it returns a tuple.

This was actually added in the previous PR that introduced the abstract attention class: https://github.com/pytorch/executorch/blob/440a3ac7800816898578d05cd2c1a76cbf787760/examples/models/llama/attention.py#L30.

cccclai commented on examples/models/llama/model.py:
```diff
@@ -289,16 +289,18 @@ def get_example_inputs_kvcache_sdpa(self):
-                    [0], dtype=torch.long
-                ),  # start_pos, what token of output are we on.
+                {
+                    "input_pos": torch.tensor(
```
Is it because we are plugging in the `Forward` option down the line?

sxu replied:
Yeah, "input_pos" is the only thing the regular attention needs, other implementations can take additional/different parameters.

cccclai reviewed (approved):
Generally seems fine

iseeyuan commented:
Thanks @sxu ! Not sure how to connect to a discussion, but this PR is related to https://github.com/pytorch/executorch/discussions/8228

sxu committed:
Pass ForwardOptions from top level module and also return any relevant state as output (#8186)

Summary:

Pass a `ForwardOptions` argument (introduced by https://github.com/pytorch/executorch/pull/8128) from the top level transformer, consolidate some existing inputs into it, and return any optional updates from the attention implementation.

Reviewed By: iseeyuan, cccclai

Differential Revision: D69080123
